runtime/fastly/host-api/host_api.cpp

```diff
@@ -10,26 +10,92 @@
 #include <algorithm>
 #include <arpa/inet.h>
 
+#include <time.h>
+
 using api::FastlyResult;
 using fastly::FastlyAPIError;
+using host_api::MonotonicClock;
 using host_api::Result;
 
+#define NEVER_HANDLE 0xFFFFFFFE
+
 size_t api::AsyncTask::select(std::vector<api::AsyncTask *> *tasks) {
   size_t tasks_len = tasks->size();
-  fastly_compute_at_edge_async_io_handle_t *handles =
-      new fastly_compute_at_edge_async_io_handle_t[tasks_len];
-  for (int i = 0; i < tasks_len; i++) {
-    handles[i] = tasks->at(i)->id();
+  std::vector<fastly_compute_at_edge_async_io_handle_t> handles;
+  handles.reserve(tasks_len);
+  uint64_t now = 0;
+  uint64_t soonest_deadline = 0;
+  size_t soonest_deadline_idx = -1;
+  for (size_t idx = 0; idx < tasks_len; ++idx) {
+    auto task = tasks->at(idx);
```

Comment:
I like `auto *` to make it obvious that the named value is a pointer vs something that implements `operator->`.

```suggestion
    auto *task = tasks->at(idx);
```